File: server/game/MapGenerator.js
const CONSTANTS = require('../../shared/constants');
const World = require('../../shared/world');
const PlayerState = require('../../shared/playerState');

// Generate initial map with players, bases, resources, and ball
class MapGenerator {
  static generate(playerIds) {
    const world = new World();
    const playerStates = new Map();

    const worldPixelWidth = CONSTANTS.MAP_WIDTH * CONSTANTS.TILE_WIDTH;
    const worldPixelHeight = CONSTANTS.MAP_HEIGHT * CONSTANTS.TILE_HEIGHT;
    const centerX = worldPixelWidth / 2;
    const centerY = worldPixelHeight / 2;

    // Add players to world
    playerIds.forEach((playerId, index) => {
      world.addPlayer(playerId, index);
      playerStates.set(playerId, new PlayerState(playerId, index));
    });

    // Create bases for each player
    const baseOffsetX = 200; // Distance from edge
    const baseY = centerY;

    playerIds.forEach((playerId, index) => {
      const baseX = index === 0 ? baseOffsetX : worldPixelWidth - baseOffsetX;
      const base = world.createActorFromDef('building', 'base', baseX, baseY, playerId);

      // Create starting workers around base
      for (let i = 0; i < CONSTANTS.STARTING_WORKERS; i++) {
        const angle = (i / CONSTANTS.STARTING_WORKERS) * Math.PI * 2;
        const workerSpawnRadius = 80; // Far enough to avoid base collision
        const workerX = baseX + Math.cos(angle) * workerSpawnRadius;
        const workerY = baseY + Math.sin(angle) * workerSpawnRadius;
        world.createActorFromDef('unit', 'worker', workerX, workerY, playerId);
      }
    });

    // Create mineral nodes
    // Near each base (easy access)
    playerIds.forEach((playerId, index) => {
      const baseX = index === 0 ? baseOffsetX : worldPixelWidth - baseOffsetX;

      // Minerals near each base
      const mineralOffset = 150;
      const mineralY1 = baseY - 150;
      const mineralY2 = baseY + 150;
      const mineralX = index === 0 ? baseX + mineralOffset : baseX - mineralOffset;

      world.createActorFromDef('resource', 'minerals', mineralX, mineralY1, null);
      world.createActorFromDef('resource', 'minerals', mineralX, mineralY2, null);
    });

    // Center minerals (contested)
    const centerMineralOffsetY = 200;
    world.createActorFromDef('resource', 'minerals', centerX, centerY - centerMineralOffsetY, null);
    world.createActorFromDef('resource', 'minerals', centerX, centerY + centerMineralOffsetY, null);

    // Create the ball in the center
    const ball = world.createActorFromDef('ball', 'ball', centerX, centerY, null);
    ball.velocityX = 0;
    ball.velocityY = 0;

    return { world, playerStates };
  }
}

module.exports = MapGenerator;

File: server/Setup/MapGenerator.cs
using server.ECS;
using server.Rooms;

namespace server.Setup;

/// <summary>
/// Generates the initial map layout for a game match.
/// Creates symmetrical map with bases, workers, mineral patches, ball, and avatars.
/// </summary>
public static class MapGenerator
{
    /// <summary>
    /// Generate the map for a match.
    /// </summary>
    /// <param name="world">The ECS World to populate</param>
    /// <param name="players">List of PlayerInfo for each player</param>
    public static void Generate(World world, IEnumerable<PlayerInfo> players)
    {
        var playerList = players.ToList();

        // Create ball at center
        CreateBall(world);

        // Create entities for each player
        foreach (var player in playerList)
        {
            CreatePlayerEntities(world, player);
        }

        // Create center mineral patches (contested resources)
        CreateCenterMinerals(world);
    }

    private static void CreateBall(World world)
    {
        var centerX = GameConstants.WORLD_PIXEL_WIDTH / 2f;
        var centerY = GameConstants.WORLD_PIXEL_HEIGHT / 2f;

        EntityFactory.CreateBall(world, centerX, centerY);
    }

    private static void CreatePlayerEntities(World world, PlayerInfo player)
    {
        var isLeftSide = player.PlayerIndex == 0;

        // Base position: 200px from edge
        var baseX = isLeftSide ? 200f : GameConstants.WORLD_PIXEL_WIDTH - 200f;
        var baseY = GameConstants.WORLD_PIXEL_HEIGHT / 2f;

        // Create base
        EntityFactory.CreateBuilding(world, "base", baseX, baseY, player.ConnectionId, player.PlayerIndex);

        // Create avatar (100px toward center from base)
        var avatarX = isLeftSide ? baseX + 100f : baseX - 100f;
        EntityFactory.CreateAvatar(world, avatarX, baseY, player.ConnectionId, player.PlayerIndex);

        // Create starting workers (4 workers around base)
        CreateStartingWorkers(world, baseX, baseY, player);

        // Create mineral patches near base (2 patches)
        CreateBaseMinerals(world, baseX, baseY, isLeftSide);
    }

    private static void CreateStartingWorkers(World world, float baseX, float baseY, PlayerInfo player)
    {
        var workerCount = GameConstants.STARTING_WORKERS;
        var spawnRadius = 80f;

        for (int i = 0; i < workerCount; i++)
        {
            var angle = (2 * Math.PI / workerCount) * i + Math.PI / 4; // Offset to avoid overlapping with base
            var workerX = baseX + (float)Math.Cos(angle) * spawnRadius;
            var workerY = baseY + (float)Math.Sin(angle) * spawnRadius;

            EntityFactory.CreateUnit(world, "worker", workerX, workerY, player.ConnectionId, player.PlayerIndex);
        }
    }

    private static void CreateBaseMinerals(World world, float baseX, float baseY, bool isLeftSide)
    {
        // Minerals are placed further from the base, toward the center
        var mineralOffset = isLeftSide ? 250f : -250f;
        var mineralX = baseX + mineralOffset;

        // Two mineral patches, one above and one below
        EntityFactory.CreateResource(world, "minerals", mineralX, baseY - 150f);
        EntityFactory.CreateResource(world, "minerals", mineralX, baseY + 150f);
    }

    private static void CreateCenterMinerals(World world)
    {
        var centerX = GameConstants.WORLD_PIXEL_WIDTH / 2f;
        var centerY = GameConstants.WORLD_PIXEL_HEIGHT / 2f;

        // Center minerals are placed diagonally from the center
        // Top-left and bottom-right of center (to be equidistant from both players)
        EntityFactory.CreateResource(world, "minerals", centerX - 300f, centerY - 400f);
        EntityFactory.CreateResource(world, "minerals", centerX + 300f, centerY + 400f);
    }
}
